conn.go

```diff
@@ -1,468 +1,47 @@
 package reliable
 
 import (
-	"fmt"
-	"github.com/lithdew/seq"
 	"io"
 	"net"
-	"sync"
-	"time"
 )
 
-type Conn struct {
-	writeBufferSize uint16 // write buffer size that must be a divisor of 65536
-	readBufferSize  uint16 // read buffer size that must be a divisor of 65536
-
-	updatePeriod  time.Duration // how often time-dependant parts of the protocol get checked
-	resendTimeout time.Duration // how long we wait until unacked packets should be resent
-
-	conn net.PacketConn
-	addr net.Addr
-	pool *Pool
-
-	ph PacketHandler
-	eh ErrorHandler
-
-	mu   sync.Mutex    // mutex over everything
-	die  bool          // is this conn closed?
-	exit chan struct{} // signal channel to close the conn
-
-	lui uint16    // last sent packet index that hasn't been sent via an ack yet
-	oui uint16    // oldest sent packet index that hasn't been acked yet
-	ouc sync.Cond // stop writes if the next write given oui may flood our peers read buffer
-	ls  time.Time // last time data was sent to our peer
+type transmitFunc func(addr net.Addr, buf []byte) error
 
-	wi uint16 // write index
-	ri uint16 // read index
-
-	wq []uint32 // write queue
-	rq []uint32 // read queue
-
-	wqe []writtenPacket // write queue entries
+type Conn struct {
+	conn     net.PacketConn
+	protocol *Protocol
 }
 
-func NewConn(conn net.PacketConn, addr net.Addr, opts ...ConnOption) *Conn {
-	c := &Conn{conn: conn, addr: addr, exit: make(chan struct{})}
-
-	for _, opt := range opts {
-		opt.applyConn(c)
-	}
-
-	if c.writeBufferSize == 0 {
-		c.writeBufferSize = DefaultWriteBufferSize
-	}
-
-	if c.readBufferSize == 0 {
-		c.readBufferSize = DefaultReadBufferSize
-	}
-
-	if c.resendTimeout == 0 {
-		c.resendTimeout = DefaultResendTimeout
-	}
-
-	if c.updatePeriod == 0 {
-		c.updatePeriod = DefaultUpdatePeriod
-	}
-
-	if c.pool == nil {
-		c.pool = new(Pool)
-	}
-
-	c.wq = make([]uint32, c.writeBufferSize)
-	c.rq = make([]uint32, c.readBufferSize)
-
-	emptyBufferIndices(c.wq)
-	emptyBufferIndices(c.rq)
-
-	c.wqe = make([]writtenPacket, c.writeBufferSize)
-
-	c.ouc.L = &c.mu
-
-	return c
+func NewConn(conn net.PacketConn, addr net.Addr, opts ...ProtocolOption) *Conn {
+	p := NewProtocol(addr, opts...)
```

Comment:
Can you have `Protocol` have its `writePacket` method return a byte slice to send, which will then be used by `Conn` to write to a `net.PacketConn`?

Reply:
This way, passing in `c.transmit` will not be necessary.

Reply:
I changed `writePacket` and `Read` return a byte slice.
I stop passing `c.transmit` as long as possible.